Remove `Array.new` and `Array.copy` and move Vector functions to builtins.

jdunkerley opened the pull request:
### Pull Request Description

- Removed Array methods: `new`, `copy` and `new_[1234]`.
- New builtins for `Vector.insert`, `Vector.remove` and `Vector.flatten`.
- Replaced `Vector_Builder` use of `Array.copy` to a `Vector.Builder` approach.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

jdunkerley committed:
Removing `Array.new_[1234]`.
Removing `Array.empty`.

jdunkerley committed:
Fix doc comment.

jdunkerley committed:
Remove the use to `to_array`.

jdunkerley committed:
Fix failing test.

jdunkerley committed:
Restore `to_array` for the builtins which can support it.

jdunkerley committed:
Fix failed test.

jdunkerley committed:
Add `remove_builtin`.

hubertp reviewed (commented):
So the idea with `Array` and `Vector` is to make sure that they have essentially the same methods and can be used interchangeably (see also #6299). It seems like the `allocate` bit interfere with that.

Re builtins design - lgtm so far.

jdunkerley commented:
> So the idea with `Array` and `Vector` is to make sure that they have essentially the same methods and can be used interchangeably (see also #6299). It seems like the `allocate` bit interfere with that.

Ideally, I will remove `allocate` and `copy` shortly. This will mean only way to make an `Array` within Enso is by creating a vector and calling `to_array`.

jdunkerley committed:
Add `flatten` as a builtin.

jdunkerley committed:
Remove `allocate` and `copy` from builtins.
Replaces the `Array.copy` in `Vector_Builder` with a new `Builder` based approach.

jdunkerley committed:
Fix Array issues using Vector builtins.

jdunkerley committed:
More fixes.

jdunkerley committed:
Java format.

radeusgd commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Array.enso:
```diff
@@ -286,7 +195,12 @@ type Array
-    remove self at=-1 = Vector.remove self at
+    remove self at=-1 = (Vector.from_polyglot_array self).remove at
+
+    ## PRIVATE
+    remove_builtin self at =
+        _ = [at]
+        Error.throw (Unimplemented.Error "Array.remove_builtin.")
```
Ideally, I think we should filter functions whose doc contains `PRIVATE` in the `VectorArraySignatureTest` and ignore them.

But I'm not sure how hard it is to do that (cc: @JaroslavTulach)

Still, I think worth adding at least a small comment inside of this function (not a doc comment obviously) explaining why it has to exist here (and hopefully a TODO to fix this at some point).

radeusgd commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Vector.enso:
```diff
@@ -717,10 +701,10 @@ type Vector a
+            self.remove_builtin used_index
+
+    ## PRIVATE
+    remove_builtin self at = @Builtin_Method("Vector.remove_builtin")
```
Let's use consistent 'calling' syntax.

```suggestion
    remove_builtin self at = @Builtin_Method "Vector.remove_builtin"
```

radeusgd commented on engine/runtime/src/test/java/org/enso/interpreter/test/ArrayTest.java:
```diff
@@ -29,7 +29,7 @@ public void dontExposeNullsOutOfArray() throws Exception {
 
     null =
-        a = Array.new 1
+        a = Array.allocate 1
```
I thought that `Array.allocate` is _not_ exposed to Enso?

radeusgd reviewed (approved):
Looks good, just a question.

jdunkerley committed:
PR comments.

jdunkerley committed:
Avoid initialising Array values.
Removed old test now allocate removed.

jdunkerley committed:
Java format.
Linting type sig check fixes.

hubertp commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Array.enso:
```diff
@@ -441,7 +359,7 @@ type Array
             [[1, 2, 3].to_array, [4, 10].to_array, [].to_array, [0].to_array, [0].to_array].to_array . flatten == [1, 2, 3, 4, 10, 0, 0].to_array
     flatten : Vector Any
-    flatten self = Vector.flatten self
+    flatten self = (Vector.from_polyglot_array self).flatten
```
That seems like a weird restriction. I'd rather we keep the call using the static wrapper and add a specialization to `flatten`, unless it is not possible.

hubertp commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Array.enso:
```diff
@@ -798,4 +727,4 @@ type Array
              [1].to_array + [2].to_array
     + : Vector Any -> Vector Any
-    + self that = Vector.+ self that
+    + self that = (Vector.from_polyglot_array self) + that
```
Similar comment as with the `flatten` change.

hubertp reviewed (commented):
I'd rather we keep some of the static wrapper calls in place, if possible

jdunkerley committed:
Adding Array support for insert.

hubertp committed:
Deal with Array-like selfs

Passing `ArraySlice` would enter the fallback and crash.

jdunkerley committed:
Adding Array support for RemoveAt.

jdunkerley committed:
Adding Array support for Flatten.

jdunkerley committed:
Revert a mistake.

jdunkerley committed:
Extra argument.

JaroslavTulach commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Array.enso:
```diff
@@ -279,6 +183,9 @@ type Array
     insert self at=self.length item=Nothing = Vector.insert self at item
 
+    ## PRIVATE
+    insert_builtin self at values = @Builtin_Method "Vector.insert_builtin"
```
What is the reason for `insert_builtin` to be defined on `Array` when nobody call it?

JaroslavTulach commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso:
```diff
@@ -473,8 +472,8 @@ Text.replace self term replacement case_sensitivity=Case_Sensitivity.Sensitive o
-                Nothing -> Array.empty
-                _ -> Array.new_1 result
+                Nothing -> []
+                _ -> [result]
```
Using `[]` and `[x]` seems really simpler than `Array.empty` and `Array.new_1`.

JaroslavTulach commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Vector.enso:
```diff
@@ -568,13 +565,7 @@ type Vector a
-            Array.copy vec.to_array 0 arr i vec.length
-            i + vec.length
-        Vector.from_polyglot_array arr
+    flatten self = @Builtin_Method "Vector.flatten"
```
I am surprised flatten isn't written in Enso and requires a builtin. Less builtins is always better...
```haskell
flat x =
    size = x.map .length . fold 0 (+)
    b = Vector.new_builder size
    x.map arr->
      arr.map e->
        b.append e
    b.to_vector
```
this code seems to `flatten` the elements too.

JaroslavTulach commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Vector.enso:
```diff
@@ -691,16 +682,13 @@ type Vector a
         _ -> Error.throw (Type_Error.Error Vector that "that")
 
+    ## PRIVATE
+    insert_builtin self at values = @Builtin_Method("Vector.insert_builtin")
```
If we made `insert_builtin` static method (here or elsewhere) we wouldn't need to define it in `Array`.

JaroslavTulach commented on distribution/lib/Standard/Base/0.0.0-dev/src/System/File.enso:
```diff
@@ -580,9 +580,9 @@ type File
-                  self.move_builtin destination copy_options
-                False -> self.move_builtin destination Array.empty
+                  copy_options = [StandardCopyOption.REPLACE_EXISTING]
+                  self.move_builtin destination copy_options.to_array
```
Nobody shall require `.to_array` call. The `File` should be fixed to work with `Vector` or any array-like object. The issue is already tracked as 
- #7133

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/immutable/FlattenVectorNode.java:
```diff
@@ -0,0 +1,92 @@
+      for (long i = 0; i < length; i++) {
+        var item = interop.readArrayElement(storage, i);
+        if (!interop.hasArrayElements(item)) {
+          Builtins builtins = EnsoContext.get(this).getBuiltins();
```
Assuming the error isn't frequent, don't include it in the compilation. Either by `CompilerDirectives.transferToInterpreter()` or by `BranchProfile.enter()`.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/immutable/FlattenVectorNode.java:
```diff
@@ -0,0 +1,92 @@
+      return Vector.fromArray(result);
+    } catch (UnsupportedMessageException | InvalidArrayIndexException e) {
+      CompilerDirectives.transferToInterpreter();
+      throw new IllegalStateException(e);
```
Don't throw `IllegalStateException`, that would crash the Enso interpreter. Throw `PanicException`.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/immutable/InsertBuiltinVectorNode.java:
```diff
@@ -0,0 +1,114 @@
+  private Vector insertBuiltin(
+      Object current, long index, Object values, CopyNode copyNode, InteropLibrary interop) {
+    try {
+      long current_length = interop.getArraySize(current);
```
Shouldn't we be using camelCase in Java?

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/immutable/RemoveAtVectorNode.java:
```diff
@@ -0,0 +1,82 @@
+      return Vector.fromArray(array);
+    } catch (UnsupportedMessageException e) {
+      CompilerDirectives.transferToInterpreter();
+      throw new IllegalStateException(e);
```
Throw `PanicException` to avoid crashing the Enso interpreter.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/mutable/CoerceArrayNode.java:
```diff
@@ -31,18 +31,8 @@ Object[] doArray(Array arr) {
-      throw new PanicException(
-          builtins.error().makeInvalidArrayIndex(arr, e.getInvalidIndex()), this);
-    }
+  Object[] doVector(Vector arr, @Cached CoerceArrayNode coerceArrayNode) {
```
Interesting change. I am not sure why it was needed? Anyway, it looks correct.

JaroslavTulach commented on engine/runtime/src/test/java/org/enso/interpreter/test/ArrayTest.java:
```diff
@@ -1,43 +0,0 @@
-    null =
-        a = Array.new 1
-        b = a.at 0
-        b
```
OK, we don't need this test.

JaroslavTulach commented on engine/runtime/src/test/scala/org/enso/interpreter/test/semantic/SystemProcessTest.scala:
```diff
@@ -131,11 +124,10 @@ class SystemProcessTest extends InterpreterTest with OsSpec {
           |
           |main =
-          |    result = System.create_process "bash" (Array.new_2 "-c" "wc -c") "" True True True
+          |    result = System.create_process "bash" ["-c", "wc -c"] "" True True True
```
Now the syntax looks like a normal syntax in normal language - way better than `Array.new_2`!

JaroslavTulach reviewed (approved):
Overall very positive and useful change. Some comments, advises left for your consideration.

It'd be good to benchmark `flatten`, `insert` and `remove` and compare previous and new results.

jdunkerley committed:
PR comments.

jdunkerley committed:
Java format.

jdunkerley committed:
Fix `Vector.+`.

jdunkerley committed:
Change `self` name for the static builtins.

jdunkerley committed:
Java format.
Some type signature fixes.
